### PR TITLE
[Core] Enhancements and refactoring of LoRA method

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Adds and improves utils for LoRA
2. Refactors LoRA for simplicity and removing redundancies.
3. Adds try catch block in PeftModel's `_prepare_model_for_gradient_checkpointing` as custom models won't have transformers specific gradient ckpt API.
4. Adds relevant docs.

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -271,15 +271,18 @@ def _prepare_model_for_gradient_checkpointing(self, model: PreTrainedModel):
...
 
-                def make_inputs_require_grad(module, input, output):
-                    output.requires_grad_(True)
+                    def make_inputs_require_grad(module, input, output):
+                        output.requires_grad_(True)
 
-                model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
+                    model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
+        except:
+            pass
```
Comment (younesbelkada) [inline src/peft/peft_model.py:285]: I don't think this is needed as the `getattr(model, "is_loaded_in_8bit", False)` already protects from the case where the model doesn't have that attribute no? I would say maybe let's avoid adding bare try / except blocks in the code base.

**Review by younesbelkada — COMMENTED** [review]
The refactoring looks nice ! I left one suggestion, also can you make sure slow tests pass just in case? 
We could also add tests for delete_adapter / unload but can be done in a follow up PR I think

File: src/peft/peft_model.py (hunk 1 above)
Comment (pacman100) [reply src/peft/peft_model.py:285]: for some reason this was leading to issues when using Auto1111 webui with PEFT which uses CompVis implementation of SD.

Comment (pacman100) [reply src/peft/peft_model.py:285]: Now I remember,  Compvis SD is when loaded in float16, will go to the `else` block and it won't have attribute `get_input_embeddings ` leading to error. @younesbelkada, WDYT?

Comment (pacman100) [reply src/peft/peft_model.py:287]: but wanted to bring this to your notice when working with non transformers models

File: docs/source/conceptual_guides/lora.mdx [hunk 2]
```diff
@@ -43,6 +43,22 @@ While LoRA is significantly smaller and faster to train, you may encounter laten
 
 This works because during training, the smaller weight matrices (*A* and *B* in the diagram above) are separate. But once training is complete, the weights can actually be merged into a new weight matrix that is identical.
 
+## Utils for LoRA
```
Comment (BenjaminBossan) [inline docs/source/conceptual_guides/lora.mdx:46]: Would it be useful to document `merge_and_unload` here is well?

File: src/peft/peft_model.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/peft_model.py:287]: Note that I made exactly that change in my PR for custom models:

https://github.com/huggingface/peft/pull/676/files#diff-95787fcc5356568929e383449ade2c1bac5da43deccab17318652b62ed171ae7R276

I agree this is better than bare `except` (and _if_ we need to use bare `except`, it should always be `except Exception` in order not to catch `KeyboardInterrupt`).

File: src/peft/tuners/lora.py [hunk 3]
```diff
@@ -411,7 +427,7 @@ def _prepare_lora_config(peft_config, model_config):
             peft_config.target_modules = TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING[model_config["model_type"]]
         return peft_config
 
-    def merge_and_unload(self):
+    def merge_and_unload(self, merge=True):
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora.py:430]: Does it make sense to have a method called `merge_and_unload` with the option to have `merge=False`?

How about moving the content of this method inside a new method `_unload_and_optionally_merge` or some similar name. Then we have:

```python
def merge_and_unload(self):
    return self._unload_and_optionally_merge(merge=True)

def unload(self):
    return self._unload_and_optionally_merge(merge=False)
```

File: src/peft/tuners/lora.py [hunk 4]
```diff
@@ -464,12 +481,37 @@ def merge_and_unload(self):
...
-        )
+    def add_weighted_adapter(self, adapters, weights, adapter_name, combination_type="svd"):
+        """
+        This method adds a new adapter by merging the given adapters with the given weights.
+
+        Args:
+            adapters (list): List of adapter names to be merged.
+            weights (list): List of weights for each adapter.
+            adapter_name (str): Name of the new adapter.
+            type (str): Type of merging. Can be one of [`svd`, `linear`]
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora.py:492]:
```suggestion
            combination_type (str): Type of merging. Can be one of [`svd`, `linear`]
```

File: src/peft/tuners/lora.py [hunk 5]
```diff
@@ -478,26 +520,85 @@ def add_weighted_adapter(self, adapters, weights, adapter_name):
...
+                        if adapter in target.lora_A:
+                            current_adapter_lora_A = target.lora_A[adapter].weight
+                            current_adapter_lora_B = target.lora_B[adapter].weight
+                        elif adapter in target.lora_embedding_A:
+                            current_adapter_lora_A = target.lora_embedding_A[adapter_name]
+                            current_adapter_lora_B = target.lora_embedding_B[adapter]
+                        target_lora_A.data += current_adapter_lora_A.data * weight * target.scaling[adapter]
+                        target_lora_B.data += current_adapter_lora_B.data * weight
+                elif combination_type == "svd":
+                    delta_weight = weights[0] * target.get_delta_weight(adapter[0])
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora.py:542]: Just a suggestion: Move this into a utility function that returns `Vh` and `U` to make this already long method a bit shorter.

File: src/peft/tuners/lora.py [hunk 6]
```diff
@@ -478,26 +520,85 @@ def add_weighted_adapter(self, adapters, weights, adapter_name):
...
+                    "r",
+                    "lora_alpha",
+                    "scaling",
+                    "lora_A",
+                    "lora_B",
+                    "lora_embedding_A",
+                    "lora_embedding_B",
+                    "lora_dropout",
+                ]:
+                    if adapter_name in getattr(target, attr):
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora.py:592]: Just wondering: Could a valid adapter name be a substring of another valid adapter name? If yes, could this `in` check result in false positives?

**Review by BenjaminBossan — COMMENTED** [review]
Very nice addition. This should give users much better control about how they deal with the adapters. I like it a lot.

I left a few comments, please take a look. I would also like to see tests for the new functionality, but I would be okay with them being added in a later PR.

File: src/peft/peft_model.py (hunk 1 above)
Comment (pacman100) [reply src/peft/peft_model.py:287]: Cool, will remove this fix as you are already doing it. Also, thanks for fixing the config part too as I was setting a dummy config manually for my usecase.

File: docs/source/conceptual_guides/lora.mdx (hunk 2 above)
Comment (pacman100) [reply docs/source/conceptual_guides/lora.mdx:46]: It is explained just above this

File: src/peft/tuners/lora.py (hunk 3 above)
Comment (pacman100) [reply src/peft/tuners/lora.py:430]: makes sense

File: src/peft/tuners/lora.py (hunk 6 above)
Comment (pacman100) [reply src/peft/tuners/lora.py:592]: Hello, here we are checking in element is present in a list rather than if a string is a substring of another string. So, this should only match if the adapter_name is exactly present in the dict keys.

File: docs/source/conceptual_guides/lora.mdx (hunk 2 above)
Comment (BenjaminBossan) [reply docs/source/conceptual_guides/lora.mdx:46]: I see, thanks

File: src/peft/tuners/lora.py (hunk 6 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora.py:592]: I see, makes sense.

File: src/peft/tuners/lora.py (hunk 3 above)
Comment (pacman100) [reply src/peft/tuners/lora.py]: Done

File: src/peft/tuners/lora.py (hunk 5 above)
Comment (pacman100) [reply src/peft/tuners/lora.py]: done

**Comment by BenjaminBossan** [comment]
@pacman100 Nice additions, especially the tests. Some are still failing, could you please check?

**Review by BenjaminBossan — APPROVED** [review]
Outstanding work, LGTM

**Review by younesbelkada — APPROVED** [review]
This looks great to me, amazing work !
